File: 459/1.py
def calc_next(s: str) -> list[int]:
    next = [-1] * len(s)
    j = 0
    for i in range(1, len(s)):
        if s[i] == s[j]:
            next[i] = next[j]
        else:
            next[i] = j
            while True:
                j = next[j]
                if j == -1 or s[j] == s[i]:
                    break
        j += 1
    return j


class Solution:
    def repeatedSubstringPattern(self, s: str) -> bool:
        j = calc_next(s)
        r = len(s) - j
        return len(s) % r == 0 and j != 0


sln = Solution()
print(sln.repeatedSubstringPattern("babbabbabbabbab"))

File: 24/1.cpp
#include <algorithm>
#include <iostream>

struct ListNode
{
  int val;
  ListNode* next;
  ListNode()
    : val(0)
    , next(nullptr)
  {}
  ListNode(int x)
    : val(x)
    , next(nullptr)
  {}
  ListNode(int x, ListNode* next)
    : val(x)
    , next(next)
  {}
};

class Solution
{
public:
  ListNode* swapPairs(ListNode* head)
  {
    auto a = head;
    if (!a)
      return nullptr;

    auto b = a->next;
    if (!b)
      return a;

    a->next = swapPairs(b->next);
    b->next = a;
    return b;
  }
};

int
main()
{
  Solution sln;
  auto head =
    new ListNode(1, new ListNode(2, new ListNode(3, new ListNode(4))));
  head = sln.swapPairs(head);

  while (head) {
    std::cout << head->val << std::endl;
    head = head->next;
  }

  getchar();
}
